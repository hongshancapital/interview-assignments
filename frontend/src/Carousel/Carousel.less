.viewBox {
  height: 100vh;
  position: relative;
  overflow: hidden;
  .innerBox {
    height: 100%;
    transition: all 0.4s;
    position: relative;

    display: flex;
    flex-direction: row;
    .imgItem {
      height: 100%;
      width: 100%;
      flex-shrink: 0;
      .textBox {
        color: red;
        height: 60%;
        display: flex;
        flex-direction: column;
        justify-content: center;
        .bigTitle {
          font-size: 32px;
          font-weight: 500;
        }
        .smallTitle {
          font-size: 20px;
        }
        .colorBlack {
          color: black;
        }
        .colorWhite {
          color: white;
        }
      }
    }
    .iphoneItem {
      background-image: url("../assets/iphone.png");
      background-size: 100% 100%;
      background-repeat: no-repeat;
    }
    .tabletItem {
      background-image: url("../assets/tablet.png");
      background-size: 100% 100%;
      background-repeat: no-repeat;
    }
    .airpodsItem {
      background: url("../assets/airpods.png") no-repeat center/cover;
    }
  }
  .btnBox {
    position: absolute;
    bottom: 0;
    height: 5%;
    width: 100%;
    text-align: center;
    .btnItem {
      display: inline-block;
      width: 80px;
      //padding: 8px;
      height: 6px;
      margin-right: 6px;
      display: inline-block;
      cursor: pointer;
      border-radius: 4px;
      background-color: gray;
      position: relative;
      .inner {
        position: absolute;
        height: 100%;
        width: 0;
        left: 0;
        border-radius: 4px;
        background: white;
      }
      &.activeBtn .inner {
        animation: widthGrow 2s linear;
      }
      @keyframes widthGrow {
        0% {
          width: 0;
        }
        100% {
          width: 100%;
        }
      }
    }
  }
}
