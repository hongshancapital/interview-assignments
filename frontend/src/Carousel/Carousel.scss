.carousel {
  position: relative;
  height: 100%;
  width: 100%;

  .carousel-items {
    position: relative;
    width: 100%;
    height: 100%;
    overflow: hidden;

    .carousel-item {
      position: absolute;
      width: 100%;
      height: 100%;
    }
  }

  .carousel-dots {
    position: absolute;
    bottom: 24px;
    width: 100%;
    display: flex;
    justify-content: center;

    .carousel-dot {
      display: inline-block;
      border-radius: 2px;
      height: 4px;
      width: 48px;
      background-color: rgba(153, 153, 153, 0.8);
      margin-right: 8px;

      &:last-child {
        margin-right: 0;
      }

      .carousel-dot-process {
        width: 100%;
        height: 100%;
        border-radius: 2px;
        background-color: rgba(255, 255, 255, 0.4);
      }
    }
  }

}