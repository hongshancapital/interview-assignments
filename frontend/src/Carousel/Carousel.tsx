import { FC } from 'react';
import { APP_CAROUSEL_ITEMS, DEFAULT_CAROUSEL_DURATION } from './constants';
import { CarouselItem, CarouselItemProps } from './CarouselItem';
import { Indicators } from './Indicators';
import styles from './Carousel.module.scss';
import { useIndex } from './hooks';

type CarouselProps = {
  /**
   * Time to display each item
   * @default 3000ms
   */
  duration?: number;
  carouselItems?: CarouselItemProps[];
};

/**
 * Carousel can display items for given duration(default to 3000ms)
 * only support fullscreen slide for now
 */
export const Carousel: FC<CarouselProps> = ({
  duration = DEFAULT_CAROUSEL_DURATION,
  carouselItems = APP_CAROUSEL_ITEMS,
}) => {
  const carouselItemsLength = carouselItems.length;
  const { currIndex } = useIndex({
    duration,
    childrenLength: carouselItemsLength,
  });

  return (
    <div className={styles.carousel}>
      <ul
        className={styles.carouselItems}
        style={{
          width: `${carouselItemsLength * 100}vw`,
          transform: `translateX(-${currIndex * 100}vw)`,
        }}
        aria-label="carousel items"
      >
        {carouselItems.map((carouselItemProps) => {
          return (
            <li
              className={styles.carouselItemWrapper}
              key={carouselItemProps.title}
            >
              <CarouselItem {...carouselItemProps} />
            </li>
          );
        })}
      </ul>
      <Indicators
        length={carouselItemsLength}
        currentActiveIndex={currIndex}
        duration={duration}
      />
    </div>
  );
};
