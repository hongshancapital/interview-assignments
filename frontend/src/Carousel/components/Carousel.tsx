import { forwardRef, useImperativeHandle, useMemo, useRef } from 'react';
import useCarousel from '../hooks/useCarousel';
import { IUseCarouselOptions } from '../types';

import CarouselItem from './CarouselItem';
import CarouselDot from './CarouseDot';

import '../style/index.scss';


/** React Component Carousel ref */
interface ICarouselRef {
  /** 跳转到指定step */
  goTo: (key: string) => void;
  next: () => void;
  last: () => void;
}

interface ICarouselProps extends IUseCarouselOptions {
  className?: string;
}

const Carousel = forwardRef<ICarouselRef, ICarouselProps>((props, ref) => {
  const { items, defaultActiveKey, autoPlay = false, autoPlayGap } = props;

  const containerRef = useRef<HTMLDivElement>(null);

  const { activeKey, itemWidth, itemListStyle, goTo, next, last } = useCarousel({
    items,
    autoPlay,
    defaultActiveKey,
    autoPlayGap,
    containerRef,
  })

  useImperativeHandle(ref, () => ({
    goTo,
    next,
    last,
  }), [goTo, next, last])

  const className = useMemo(() => {
    const classNames = ['carousel'];
    if (typeof props.className === 'string') classNames.push(props.className);
    return classNames.join(' ');
  }, [props.className])

  // 第一次渲染，itemWidth 不存在, 不用进行卡片的渲染
  if (itemWidth === undefined) return <div ref={containerRef} className={className} />;

  return (
    <div ref={containerRef} className={className}>
      <div className='carousel-dot-wrapper'>
        <div className='carousel-dot-list'>
          {items.map(item => (
            <CarouselDot onDotClick={() => goTo(item.key)} key={item.key} active={activeKey === item.key && autoPlay} selected={activeKey === item.key && !autoPlay}/>
          ))}
        </div>
      </div>

      <div style={itemListStyle} className="carousel-item-list">
        {items.map(item => (
          <CarouselItem key={item.key} width={itemWidth}>{item.children}</CarouselItem>
        ))}
      </div>
    </div>
  )
})

if (process.env.NODE_ENV !== 'production') {
  Carousel.displayName = 'Carousel';
}

export default Carousel;