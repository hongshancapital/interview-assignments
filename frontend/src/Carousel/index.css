html, body {
    width: 100%;
    height: 100%;
}
.carousel {
    width: 100%;
    height: 100vh;
    overflow: hidden;
    position: relative;
    margin: auto;
    background-color: white;
}


.carousel-frame {
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    right: 0;
    bottom: 0;
    /* animation-name: imgMove; */
    animation-duration: 20s;
    animation-timing-function: cubic-bezier(0.075, 0.82, 0.165, 1);
    /* animation-delay: 5s; */
    /* animation-iteration-count: infinite; */
}

.carousel-frame-item {
    position: absolute;
    width: 100%;
    height: 100%;
}

.carouse-item {
    position: relative;
    width: 100%;
    height: 100%;
    display: flex;
    flex-direction: column;
}

.carouse-item img {
    position: absolute;
    left: 0;
    bottom: 0;
    width: 100%;
    height: auto;
}

.carouse-item .carouse-text-wrap {
    z-index: 500;
    margin-top: 10%;
    font-size: 24px;
}

.carouse-item .carouse-text-wrap .item-title {
    font-size: 34px;
    font-weight: bold;
}

.indicator-wrap {
    display: block;
    position: absolute;
    left: 50%;
    bottom: 10px;
    transform: translateX(-50%);
}

.indicator {
    display: inline-block;
    background-color: #787878;
    width: 20px;
    height: 2px;
    border-radius: 1px;
    overflow: hidden;
}

.indicator:not(:last-child){
    margin-right: 5px;
}

.indicator-child {
    width: 22px;
    height: 2px;
    background-color: white;
    border-radius: 1px;
    transform: translateX(-100%);
}

/* @keyframes imgMove {
    0% {
        transform: translateX(0px);
    }

    25% {
        transform: translateX(-100%);
    }

    50% {
        transform: translateX(-200%);
    }

    75% {
        transform: translateX(-300%);
    }

    100% {
        transform: translateX(-400%);
    }
} */