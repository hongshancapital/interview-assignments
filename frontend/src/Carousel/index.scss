.carousel {
    position: relative;
    width: 100vw;
    height: 100vh;
}