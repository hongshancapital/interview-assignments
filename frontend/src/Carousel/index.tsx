import React, { ReactElement, useCallback, useEffect, useRef, useState } from "react";
import './index.css'

interface CarouselProps {
    carouseItemList: Array<ReactElement>, // 子元素列表
    duration: number, // 动画时长
    delay:number // 延迟时长
}

export default function Carousel(props: CarouselProps) {

    const { carouseItemList, duration, delay } = props

    const [position, setPosition] = useState(0)

    // 需要执行动画组件的引用
    const carouselFrameRef = useRef(null)

    // 指示器引用
    const indicatorWrapRef = useRef(null)

    // 指示器执行动画
    const translateIndicator = useCallback(()=> {
        if(indicatorWrapRef.current) {
            const indicatorWrap = indicatorWrapRef.current as HTMLElement
            indicatorWrap.childNodes.forEach((indicator, index)=>{
                const indicatorChild = indicator.firstChild as HTMLElement
                indicatorChild.style.transitionDelay = '0s'
                if(index === position) {
                    indicatorChild.style.transform = 'translateX(0%)'
                    indicatorChild.style.transitionDuration = `${(delay+duration)/1000}s`
                    indicatorChild.style.transitionTimingFunction = "ease"
                } else {
                    indicatorChild.style.transform = 'translateX(-100%)'
                    indicatorChild.style.transitionDuration = "0s"
                }
            })
        }
    },[delay,duration, position])

    // item执行轮播动画
    const translateCarouselItem = useCallback(()=>{
        if(carouselFrameRef.current) {
            const ele = carouselFrameRef.current as HTMLElement
            const tx = position * 100
            ele.style.transform = `translateX(-${tx}%)`
        }
    },[position])

    // timeout引用
    const timeoutRef = useRef(setTimeout(()=>{}))

    // 当position发生变化，执行轮播动画
    useEffect(()=>{
        translateCarouselItem()
        translateIndicator()
        let nextP = position + 1
        if(nextP >= carouseItemList.length) {
            nextP = 0
        }
        if(timeoutRef.current) {
            clearTimeout(timeoutRef.current)
        }
        timeoutRef.current = setTimeout(()=>{
            setPosition(nextP)
        }, duration + delay)
        return ()=>{
            if(timeoutRef.current) {
                clearTimeout(timeoutRef.current)
            }
        }
    },[position, carouseItemList, delay, duration, translateCarouselItem, translateIndicator])

    // carouseFrame样式
    const carouselFrameStyle = {
        transitionDuration: `${duration/1000}s`,
        transitionTimingFunction: "ease",
    }

    return <div className="carousel">
        <div ref={carouselFrameRef} className="carousel-frame" style={carouselFrameStyle}>
            {/* 轮播图 */}
        {carouseItemList.map((item, index) => {
            const translatex = index * 100
            const styles = {transform: `translate(${translatex}%, 0px)`}
            return <div className="carousel-frame-item" style={styles} key={index}>{item}</div>
        })}
        </div>
        {/* 指示器 */}
        <div className="indicator-wrap" ref={indicatorWrapRef}>
            {carouseItemList.map((_, index) => {
                return <div key={index} className="indicator">
                    <div className="indicator-child"/>
                </div>
            })}
        </div>
    </div>
}
