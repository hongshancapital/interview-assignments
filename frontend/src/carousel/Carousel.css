.carousel {
  width: 100%;
  height: 100%;
  color: #fff;
  overflow: hidden;
  position: relative;
}
.carousel .scroll {
  height: 100%;
  overflow: hidden;
  display: flex;
}
.carousel .scroll .scroll_item {
  flex: 1;
}
.carousel .slide {
  position: absolute;
  left: 50%;
  transform: translate(-50%);
  bottom: 50px;
  display: flex;
}
.carousel .slide .slide_item {
  height: 2px;
  width: 50px;
  margin: 0 10px;
  border-radius: 6px;
  background: #ababab;
  overflow: hidden;
}
.carousel .slide .slide_item .slide_item_progress {
  background-color: #fff;
  border-radius: 6px;
  height: 100%;
  width: 0;
}
