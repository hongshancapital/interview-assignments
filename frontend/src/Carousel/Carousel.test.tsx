import { render, screen } from '@testing-library/react';
import { Carousel } from './Carousel';
import { DEFAULT_CAROUSEL_DURATION } from './constants';
import { act } from 'react-dom/test-utils';

describe('Carousel', () => {
  jest.useFakeTimers();
  it('renders correctly', () => {
    const { asFragment } = render(<Carousel />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('sets the correct transform for active item for default props', () => {
    render(<Carousel />);
    act(() => {
      jest.advanceTimersByTime(DEFAULT_CAROUSEL_DURATION);
    });
    expect(screen.getByLabelText('carousel items')).toHaveStyle({
      width: '300vw',
      transform: 'translateX(-100vw)',
    });
  });
});
