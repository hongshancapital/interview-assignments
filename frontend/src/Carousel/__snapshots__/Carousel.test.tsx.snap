// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders correctly 1`] = `
<DocumentFragment>
  <div
    class="carousel"
  >
    <ul
      aria-label="carousel items"
      class="carouselItems"
      style="width: 300vw; transform: translateX(-0vw);"
    >
      <li
        class="carouselItemWrapper"
      >
        <div
          aria-label="carousel item"
          class="carouselItem carouselItemIphone"
        >
          <h2
            class="carouselItemTitle"
          >
            xPhone
          </h2>
          <h3
            class="carouselItemSubTitle"
          >
            Lots to love.less to spend.
Starting as $399.
          </h3>
          <img
            alt="xPhone"
            class="carouselItemImage"
            src="iphone.png"
          />
        </div>
      </li>
      <li
        class="carouselItemWrapper"
      >
        <div
          aria-label="carousel item"
          class="carouselItem carouselItemTablet"
        >
          <h2
            class="carouselItemTitle"
          >
            Tablet
          </h2>
          <h3
            class="carouselItemSubTitle"
          >
            Just the right amount of everything.
          </h3>
          <img
            alt="Tablet"
            class="carouselItemImage"
            src="tablet.png"
          />
        </div>
      </li>
      <li
        class="carouselItemWrapper"
      >
        <div
          aria-label="carousel item"
          class="carouselItem carouselItemAirPod"
        >
          <h2
            class="carouselItemTitle"
          >
            Buy a Tablet or xPhone for college.
Get airPods.
          </h2>
          <h3
            class="carouselItemSubTitle"
          />
          <img
            alt="Buy a Tablet or xPhone for college.
Get airPods."
            class="carouselItemImage"
            src="airpods.png"
          />
        </div>
      </li>
    </ul>
    <div
      class="indicators"
    >
      <div
        aria-label="indicator"
        class="indicator"
        role="progressbar"
      >
        <span
          aria-label="indicator progress"
          class="indicatorProgress"
          style="opacity: 1; animation: progress 3000ms linear;"
        />
      </div>
      <div
        aria-label="indicator"
        class="indicator"
        role="progressbar"
      >
        <span
          aria-label="indicator progress"
          class="indicatorProgress"
          style="opacity: 0; animation: none;"
        />
      </div>
      <div
        aria-label="indicator"
        class="indicator"
        role="progressbar"
      >
        <span
          aria-label="indicator progress"
          class="indicatorProgress"
          style="opacity: 0; animation: none;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
